truffle/com.oracle.truffle.api.instrumentation.test/src/com/oracle/truffle/api/instrumentation/InstrumentationTest.java

```diff
@@ -667,4 +669,94 @@ public void onReturnValue(EventContext context, VirtualFrame frame, Object resul
         }
     }
 
+    /*
+     * Tests for debugger or any other clients that cancel execution while halted
+     */
+
+    @Test
+    public void testKillExceptionOnEnter() throws IOException {
+        engine.getInstruments().get("testKillQuitException").setEnabled(true);
+        TestKillQuitException.exceptionOnEnter = new KillException();
+        TestKillQuitException.exceptionOnReturnValue = null;
+        TestKillQuitException.returnExceptionalCount = 0;
+        try {
+            run("STATEMENT");
+            Assert.fail("KillException in onEnter() cancels engine execution");
+        } catch (KillException ex) {
+        }
+        Assert.assertEquals("KillException is not an execution event", 0, TestKillQuitException.returnExceptionalCount);
+    }
+
+    @Test
+    public void testQuitExceptionOnEnter() throws IOException {
+        engine.getInstruments().get("testKillQuitException").setEnabled(true);
+        TestKillQuitException.exceptionOnEnter = new QuitException();
+        TestKillQuitException.exceptionOnReturnValue = null;
+        TestKillQuitException.returnExceptionalCount = 0;
+        try {
+            run("STATEMENT");
```

Comment:
CH6) It seems that if we throw the KillExcepion all through PolyglotEngine, we need to adapt either the documentation in PolyglotEngine. (Which is an incompatible change)
I'd prefer if we wrap the KillException into an IOException inside PolyglotEngine.